components/dialog/dialog-mixin.js

```diff
@@ -177,10 +177,33 @@ export const DialogMixin = superclass => class extends RtlMixin(superclass) {
 		}
 	}
 
+	_findAutofocusElement(node) {
+		if (this._useNative) {
+			// Do not override native autofocus attribute implementation
+			return null;
+		}
+
+		if (!node.querySelector('slot')) {
+			// We are in a confirm dialog, autofocus attribute will never be set
+			return null;
+		}
+
+		const content = node.querySelector('slot').assignedElements({ flatten: true });
+		const el = content.find(el => el.getAttribute('autofocus') === '' || el.getAttribute('autofocus') === 'true')
+			?? content.find(el => el.querySelector('[autofocus]'))?.querySelector('[autofocus]');
```

Comment:
This is messy. I am open to suggestions in how to clean it up.

Reply:
Yeah this is a little awkward.  `find` doesn't seem to be the right tool for the job here. `filter` would work, but we want to stop searching once we find the element.

My first inclination is to do something like this, as it seems easy to follow, stops as soon as it finds the element, and eliminates the double query.  But yeah, if you can get something nice working for `reduce`, it may be cleaner.

``` javascript
let el;
for (let i = 0; i < content.length; i++) {
    if (content[i].hasAttribute('autofocus')) {
        el = content[i];
        break;
    }
    const temp = content[i].querySelector('[autofocus]');
    if (temp) {
        el = temp;
        break;
    }
}
```